prototypes/4.ai-scheduling/src/Scheduling/Scheduler.gd

```diff
@@ -0,0 +1,159 @@
+extends Node
+"""
+The central node that keeps track of running ongoing jobs and allocates a budget of time that any 
+job should respect.
+
+Ideally put in as an auto-loaded global node, the Scheduler is loaded with SchedulableJob reference
+objects. Every frame, it will run through its list and check if it should be updating a given job
+based on its frequency and a phase offset.
+
+The phase of a job is to prevent spikes of activity; if there are 100 AI entities created at the
+same time, all vying for an update every 10 frames, there are 9 frames where nothing happens,
+and one frame where every entity fights for time budget. The phase will offset them to try and
+spread the load. However, the phase offset can only be as big as the largest frequency-1.
+
+Higher priority jobs added via `add_new_priority_job` will get the lion's share of the budget
+before lower priority ones. In addition, there is a balanced sub scheduler for jobs added via
+`add_new_job`. It is a priority 1 job, and it will spread its allocated part of the budget evenly.
+"""
+
+#int64_t's upper maximum, an equivalent to +infinity
+const MAX_INT: = 0x7FFFFFFFFFFFFFFF
+
+#4 milliseconds maximum budget by default
+var microseconds_budget: = 4000 setget set_microseconds_budget, get_microseconds_budget
+
+var _sub_scheduler: = BalancedSubScheduler.new()
+var _current_frame: = 0
+var _simulation_frame_count: = 100
+var _job_list: = []
+var _run_list: = []
+var _phase_counters: = []
+
+
+func _ready() -> void:
+	add_new_priority_job(_sub_scheduler, 1, 1)
+
+
+# warning-ignore:unused_argument
+func _process(delta) -> void:
+	var last_time: = OS.get_ticks_usec()
+	
+	_current_frame += 1
+	
+	_run_list.clear()
+	var total_priority: float = 0
+	
+	var i: = _job_list.size()-1
+	while i >= 0:
+		var job: = ((_job_list[i] as WeakRef).get_ref() as SchedulableJob)
```

Comment:
I do it when I can (when I check that the WeakRef is still valid) but in places where job is returned as a Variant by the for loop, I can't do it as easily. I'd either have to cast it into its own type-aware variable (var casted_job: = (job as SchedulableJob) ) or change to a range(0, size) function and grab it pre-casted then.

I guess I went with the way of fewer lines and fewer variables. An easy fix to change, though.

Reply:
This ensures that the code is always type-safe, right? I haven't used casting a lot in GDScript, and it was mostly to get autocompletion in the editor. I just wasn't sure about what it did.

Reply:
In the case of getting its variables (_priority, _frequency, etc) then it doesn't need casting, to be honest, I mostly did it for autocompletion.
But in the case of calling _run(), it does need to be casted - the editor throws an error because it's not aware of _run as a function and its return value being -> int, so it'd need to be done through explicit typing, or allow type inference via casting.